### DISCUSSION TITLE
decreasing efficiency (using operating points and unit_incremental_heat_rate)

### DISCUSSION THREAD

**Discussion by hannesfelipe** [spine-tools/SpineOpt.jl#962] [Support]
### Describe your problem

Dear all, 

thank you for the feedback to model hydrogen production from electrolysis in the user meeting last Tuesday. We looked into our model and using now a combination of a minimum_operating_point, operating_points, unit_incremental_heat_rate, and ordered_unit_flow_op.  

The relationship is unit_from_node (Electrolyzer ǀ Power) and the unit_node_node relationship for the unit_incremental_heat_rate Electrolyzer ǀ Power ǀ Hydrogen. 

We tested the approach with an exemplary three-part function for the operating_points ([0.4, 0.7, 1.0]) and an unit_incremental_heat_rate ([1.47, 1.59, 1.72]). minimum_operating_point is 0.2 and ordered_unit_flow_op True. 

**The problem is:** The model is not using the correct unit_incremental_heat_rate in the different parts of the operating_points. Even though is is running in the part between 0.4 and 0.7 there is production in the part 0.7-1-0. The same happens if production is within the range 0.7 and 1.0. It is not using the available capacity for the lower operating points with higher efficiency as much as possible. 

This might be the case as the model is only using the best efficiency out of the unit_incremental_heat_rate array. Thus it is using 0.68 for all parts of the piece wise linear function provided by the operating_points. 

Can we change something or is there a better approach to address this issue?  

### What versions are you using?

Spine.Opt: 2.1.15
SpineInterface: 0.11.15
SpineToolbox: 0.7.2
Julia: 1.9.4

**Comment by DillonJ** [comment]
Hi @hannesfelipe 

Have you tried using `ordered_unit_flow_op`? https://spine-tools.github.io/SpineOpt.jl/latest/concept_reference/Parameters/#ordered_unit_flow_op

Without this, we rely on convexity to achieve the correct ordering. If you have a non convex heat rate curve, you need to use `ordered_unit_flow_op`

**Reply by hannesfelipe** [reply]
Hi @DillonJ 

sorry, seems that I forgot to say this. We have set ordered_unit_flow_op to True. or does it have to be passed as an array with several boolean values, e.g., [True, True, True]?

**Comment by hannesfelipe** [comment]
@DillonJ sorry for coming back to this topic that late. I replicate our problem in a toy model. I hope that helps to identify why it is not working in our case. 

The model is only using the first entry of the array and it seems to ignore that we have defined (in this case) three pieces of the function for the unit_incremental_heat rate. 
<img width="827" alt="simple_model_operating_points" src="https://github.com/spine-tools/SpineOpt.jl/assets/36731033/77730ad1-14d9-4990-9932-046bc0f550df">

I can also send you the model in the case that this would be easier to help us.

**Reply by nnhjy** [reply]
Hi @hannesfelipe, could you upload the case model? It would indeed help find the problem.

**Reply by hannesfelipe** [reply]
Hi @nnhjy, I never uploaded a model. Can you tell me which files you need?

**Reply by nnhjy** [reply]
Sure, just upload a zip file of the .sqlite input file would be good. Thanks!

**Reply by hannesfelipe** [reply]
Perfect. Please let me know if you need more. 

[Input.zip](https://github.com/spine-tools/SpineOpt.jl/files/15109793/Input.zip)

**Comment by hannesfelipe** [comment]
I forgot to share the results. As you can see, the efficiency is the same in all hours. There are no changes in the overall efficiency despite the changes of the demand. 

<img width="355" alt="simple_model_operating_points_result" src="https://github.com/spine-tools/SpineOpt.jl/assets/36731033/b13d7a44-ceb8-4a7d-aea1-2e0c48fcd779">

**Comment by DillonJ** [comment]
@nnhjy do you have any idea what is going on here?

**Reply by nnhjy** [reply]
I will give a try

**Comment by hannesfelipe** [comment]
@DillonJ, @nnhjy thank you for all the answers already. I think this is a bug in the unit from node relation. 

I changed all parameters to the unit to node relation and then it works as expected. I attached the input db of this version. Thus, you can compare both approaches and I hope it helps to find the issue. 

[Input_on_to_node_relation.zip](https://github.com/spine-tools/SpineOpt.jl/files/15111547/Input_on_to_node_relation.zip)

**Reply by nnhjy** [reply]
I hope this answer will help. The "to_node" approach seems the only right way to use `unit_incremental_heat_rate` as defined in [the formulation](https://github.com/spine-tools/SpineOpt.jl/blob/8e7b67dd5f1f751299fd67cece47b3d5e563c2de/src/constraints/constraint_unit_pw_heat_rate.jl#L71), where only the flow segments defined under `unit__to_node` relationship is considered. I suppose it is so designed to be in line with the very meaning of "heat rate", i.e. the quantity of heat to deliver per unit of electricity. (@DillonJ Correct me if I'm wrong. It seems that the `unit_incremental_heat_rate` appears in several places in the documentation, but there is no reminder or warning about this.)

@hannesfelipe where did you see and learn this approach? We will need to give a more clear hint in this regard.

Nevertheless, the "from_node" approach is indeed doable by using the parameter `unit_flow_coefficient` of the `user_constraint` object. Please check out this db: [piece-wise-linear-from_node.zip](https://github.com/spine-tools/SpineOpt.jl/files/15120092/piece-wise-linear-from_node.zip). Note that in this way, the `Array` coefficients for individual flow segments act as "efficiency" of converting the "fuel" into "electricity", i.e. reciprocal of heat rate.

**Reply by hannesfelipe** [reply]
Thank you for the explanations. The problem is, that a unit_incremental_heat_rate with a definition of a plain value is working for the unit__from_node relations. Thus, I think it would be beneficial to enable the operating points also for this relation (and not only for the unit__to_node) relation. I get your point of the origins of the parameter as a "heat rate" but the parameter can be used for more than just this specific case (e.g. also fuel to power - I think that was also used this way in one of the tutorials). 

I cannot say exactly where I saw the implementation of the unit_incremental_heat_rate. Potentially in a tutorial and also in the documentation and parameter definitions. As there was no warning that the parameter is not working in combination with operation points for the unit__from_node relation, I started to use it for this relation. 

Maybe some background information: We model electrolysis and other power-to-X processes. The capacity of a unit is often provided regarding the input power. From a user perspective, it is intuitive to use all the definitions for the unit__from_node. Thus, as the plain parameter is working for the unit__to_node relation, I would recommend to enable the use of variable efficiencies for both sides of the unit and node relations. I think, users would benefit from this implementation instead of setting warnings that the function can only be used for the unit__to_node relation.

**Reply by nnhjy** [reply]
Thank you for the valued feedback! As shown in the database I uploaded, both the "to_node" and "from_node" flows can have a piecewise efficiency. In that way of using the `user_constraint`, the `unit_flow_coefficient` is equivalent to the `unit_incremental_heat_rate`. 

For cases like electrolysers, which I'm/will also be working on, I would use this generic approach. It is said generic also in terms of being applicable to cases where the "efficiency" ("heat rate") is incremental (decremental) as the operating level goes up, whereas the `unit_incremental_heat_rate` approach only deals with decremental (incremental) "efficiency" ("heat rate"). And actually, this specific approach is built on the generic approach and serves as a kinda shortcut for conventional units.

But I see your point. Indeed, it is quite confusing that `unit_incremental_heat_rate` works in both flow directions with a plain value but goes wrong in one direction when defined with segmental values. We will definitely need to make it more straightforward and intuitive.

**Reply by hannesfelipe** [reply]
I have seen it in the database. Thank you for providing it. I think we will change our modelling approach now.
